File: 0419-battleships-in-a-board/0419-battleships-in-a-board.java
class Solution {
    int ROWS;
    int COLS;

    public int countBattleships(char[][] board) {
        ROWS = board.length;
        if (ROWS == 0) {
            return 0;
        }
        COLS = board[0].length;
        int count = 0;
        for (int row = 0; row < ROWS; row++) {
            for (int col = 0; col < COLS; col++) {
                if (board[row][col] == 'X') {
                    count++;
                    dfs(board, row, col);
                }
            }
        }
        return count;
    }

    public void dfs(char[][] board, int row, int col) {
        if (row < 0 || row >= ROWS || col < 0 || col >= COLS || board[row][col] == '.') {
            return;
        }
        board[row][col] = '.';

        dfs(board, row-1, col);
        dfs(board, row+1, col);
        dfs(board, row, col-1);
        dfs(board, row, col+1);
    }
}

File: 0419-battleships-in-a-board/0419-battleships-in-a-board.go
func countBattleships(board [][]byte) int {
    rows := len(board)
    if rows == 0 {
        return 0
    }
    cols := len(board[0])

    count := 0

    var dfs func(int, int)
    dfs = func(row, col int) {
        if row < 0 || row >= rows || col < 0 || col >= cols || board[row][col] == '.' {
            return
        }

        board[row][col] = '.'

        dfs(row+1, col)
        dfs(row-1, col)
        dfs(row, col+1)
        dfs(row, col-1)
    }

    for i := 0; i < rows; i++ {
        for j := 0; j < cols; j++ {
            if board[i][j] == 'X' {
                count++
                dfs(i, j)
            }
        }
    }

    return count
}
